ScoreSaber/Core/Services/ScoreSaberRichPresenceService.cs

```diff
@@ -0,0 +1,358 @@
+﻿using System;
+using HarmonyLib;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json;
+using Phoenix;
+using SiraUtil.Affinity;
+using SiraUtil.Logging;
+using WebSocketSharp;
+using Zenject;
+using System.Net.Sockets;
+using Socket = Phoenix.Socket;
+using System.Reflection;
+using System.Threading.Tasks;
+using System.Linq;
+
+namespace ScoreSaber.Core.Services {
+    internal class ScoreSaberRichPresenceService : IDisposable {
+        [Inject] private readonly SiraLog _log = null;
+        [Inject] private readonly PlayerService _playerService = null;
+
+        private Socket _socket;
+        private bool _isDisposed;
+        private bool _isEnabled = true;
+        private Channel _userProfileChannel;
+
+        private const int MaxRetries = 5;
+        private int _retryAttempt = 0;
+
+        public string TimeRightNow => DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss");
+
+        public void Initialize() {
+            if (_socket != null) {
+                _log.Warn("Socket already initialized.");
+                return;
+            }
+
+            DisconnectSocket();
+
+            if (Plugin.Settings.enableRichPresence == false) {
+                _log.Warn("Rich presence is disabled in settings.");
+                return;
+            }
+
+            try {
+                var socketOptions = new Socket.Options(new JsonMessageSerializer());
+                var socketAddress = "wss://ssrt.bzaz.au/socket"; // change to scoresaber subdomain once ready.
```

Comment:
`wss://realtime.scoresaber.com/socket`

Should be a constant defined outside of the method, should build off of the baseUrl defined `Plugin.cs` however [this](https://github.com/ScoreSaber/scoresaber-plugin/pull/30/files#r1884787395) was implemented

Reply:
i made it a constant within the file, building off of the baseUrl would be difficult because its the api url, where this is the socket

Reply:
The baseUrl should look like `scoresaber.com`, everything else should build off of this base.

Reply:
implemented

Reply:
Wouldn't it make more sense to have not a single baseUrl, but instead multiple baseUrls for the different parts (still all stored at the same spot). That way it would be much easier to test individual component changes locally without having to deploy a full blown backend with all parts/components each time.